std/assembly/util/uri.ts

```diff
@@ -0,0 +1,260 @@
+import { E_URI_MALFORMED } from "./error";
+import { CharCode } from "./string";
+
+// @ts-ignore: decorator
+@lazy export const URI_UNSAFE = memory.data<u8>([
+/* skip 32 + 1 always set to '1' head slots
+  */ 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0,
+  1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, /*
+  skip 128 + 1 always set to '1' tail slots */
+]);
+
+// @ts-ignore: decorator
+@lazy export const URL_UNSAFE = memory.data<u8>([
+/* skip 32 + 1 always set to '1' head slots
+  */ 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1,
+  1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0,
+  1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, /*
+  skip 128 + 1 always set to '1' tail slots */
+]);
+
+// @ts-ignore: decorator
+@lazy export const URI_RESERVED = memory.data<u8>([
+  /*  skip 32 + 3 always set to '0' head slots
+        */ 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1,
+  1, /* skip 191 always set to '0' tail slots */
+]);
+
+export function encode(dst: usize, src: usize, len: usize, table: usize): usize {
+  var i: usize = 0, offset: usize = 0, outSize = len << 1;
+
+  while (i < len) {
+    let org = i;
+    let c: u32, c1: u32;
+    do {
+      c = <u32>load<u16>(src + (i << 1));
+      if (c - 33 < 94) { // 127 - 33
+        if (load<u8>(table + (c - 33))) break;
+      } else break;
+    } while (++i < len);
+
+    if (i > org) {
+      let size = i - org << 1;
+      if (offset + size > outSize) {
+        outSize = offset + size;
+        dst = __renew(dst, outSize);
+      }
+      if (size == 2) {
+        store<u16>(dst + offset, load<u16>(src + (org << 1)));
+      } else if (size == 4) {
+        store<u32>(dst + offset, load<u32>(src + (org << 1)));
+      } else {
+        memory.copy(
+          dst + offset,
+          src + (org << 1),
+          size
+        );
+      }
+      offset += size;
+      if (i >= len) break;
+    }
+
+    if (c >= 0xD800) {
+      if (c >= 0xDC00 && c <= 0xDFFF) {
+        throw new URIError(E_URI_MALFORMED);
+      }
+      if (c <= 0xDBFF) {
+        if (i >= len) {
+          throw new URIError(E_URI_MALFORMED);
+        }
+        c1 = <u32>load<u16>(src + (++i << 1));
+        if (c1 < 0xDC00 || c1 > 0xDFFF) {
+          throw new URIError(E_URI_MALFORMED);
+        }
+        c = (((c & 0x3FF) << 10) | (c1 & 0x3FF)) + 0x10000;
+      }
+    }
+
+    let estSize = offset + (c < 0x80 ? 1 * 6 : 4 * 6);
+    if (estSize > outSize) {
+      outSize = len > 1 ? estSize << 1 : estSize;
```

Comment:
Can you add a comment here what this does? :)

Reply:
done!